File: core/src/main/java/com/canyon/config/ConfigProvider.java
package com.canyon.config;

import com.canyon.inject.Config;
import com.canyon.inject.InjectProvider;

public class ConfigProvider implements InjectProvider {
    @Override
    public boolean isMatch(Class<?> clazz) {
        return clazz.getAnnotation(Config.class) != null;
    }

    @Override
    public <T> T create(Class<T> objClazz, Class<T> fieldClazz) {
        Config c = fieldClazz.getAnnotation(Config.class);
        if (c.path().isEmpty()) {
            return ConfigFactory.load("application.conf", fieldClazz);
        } else {
            return ConfigFactory.load(c.path(), fieldClazz);
        }
    }
}

File: core/src/main/kotlin/com/canyon/config/ConfigProvider.kt
package com.canyon.config

import com.canyon.commons.find
import com.canyon.inject.Config
import com.canyon.inject.InjectProvider
import kotlin.reflect.KClass

class ConfigProvider : InjectProvider {

    override fun isMatch(kClass: KClass<*>): Boolean {
        return kClass.annotations.firstOrNull {
            it.annotationClass == Config::class
        } != null
    }

    override fun create(objClass: KClass<*>, fieldClass: KClass<*>): Any {
        val conf = fieldClass.annotations.find(Config::class)
        return if (conf!!.path.isEmpty()) {
            ConfigFactory.load("application.conf", fieldClass.java)
        } else {
            ConfigFactory.load(conf.path, fieldClass.java)
        }
    }
}